### PR TITLE
[speech-cmd] add speech command analysis application

### PR DESCRIPTION
**Pull request by LemonNoel**
### PR types

New features

### PR changes

Others

### Description

- Add speech command analysis application.
- Upload a screenshot of the Web demo.
d751b6f9d54e9404519f2c857a72ea01

### CODE REVIEW

File: applications/speech-cmd-analysis/README.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+针对以上场景，应用[PaddleSpeech](https://github.com/PaddlePaddle/PaddleSpeech)的语音识别技术和[PaddleNLP](https://github.com/PaddlePaddle/PaddleNLP)的信息抽取技术，可以自动识别和抽取语音中的关键信息，帮助相关人员简化记录流程，提高工作效率和质量。
+另外，通过构造小样本优化信息抽取模型，能够获得更加准确的场景定制化效果。
+
+#### 方案选型
+
+- **语音识别模型**
+  [Conformer](https://arxiv.org/abs/2005.08100): Anmol Gulati等人在2020年提出的语音识别模型，通过将卷积应用于Transfomer的Encoder层，结合了Transformer提取长序列的优势和卷积提取局部特征的优势，使得语音识别的准确率明显提高。
+
+- **信息抽取模型**
+  [UIE](https://arxiv.org/pdf/2203.12277.pdf): Yaojie Lu等人在2022年提出了开放域信息抽取的统一框架，这一框架在实体抽取、关系抽取、事件抽取、情感分析等任务上都有着良好的泛化效果。本应用基于这篇工作的prompt设计思想，提供了以ERNIE为底座的阅读理解型信息抽取模型，用于关键数据抽取。同时，针对不同场景，支持通过构造小样本数据来优化模型效果，快速适配特定的关键信息配置。
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: UIE 这块还是把全称拿出来 
关键数据->关键信息

File: applications/speech-cmd-analysis/README.md [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+- paddlepaddle >= 2.2.0
+- paddlenlp >= 2.3.0
+
+安装相关问题可参考[PaddlePaddle](https://www.paddlepaddle.org.cn/install/quick?docurl=/documentation/docs/zh/install/pip/linux-pip.html)和[PaddleNLP](https://paddlenlp.readthedocs.io/zh/latest/get_started/installation.html)文档。如需下载最新版本的PaddleNLP，可执行以下命令
+```shell
+git clone https://github.com/PaddlePaddle/PaddleNLP.git
+cd PaddleNLP
+pip install -r requirements.txt
+pip install -e .
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: 这块可能要讨论一下，git clone方式来安装会导致安装时间长，因为PaddleNLP的repo体积有100多M，拉取耗时，比较推荐pip安装包方式

File: applications/speech-cmd-analysis/README.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+cd ffmpeg
+./configure
+make
+make install
+pip install pydub
+```
+
+## 3. 数据准备
+
+本应用来自于语音报销工单信息录入场景，即从语音中抽取出报销需要的``时间``、``出发地``、``目的地``和``费用``字段。相应的数据集为[语音报销工单数据](https://paddlenlp.bj.bcebos.com/datasets/erniekit/speech-cmd-analysis/audio-expense-account.jsonl)，共50条标注数据，用于信息抽取模型在交通费报销场景下的优化，示例数据如下：
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: 这块没有说清楚场景，需要再细化一下，直接说公司打车语音报销工单，数据巍语音工单转化为文字信息

File: applications/speech-cmd-analysis/README.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+- ``audio_file``: 原始音频文件或者所在目录。若设置为目录，则对该目录下所有音频文件进行格式转换。
+- ``audio_format``: 原始音频文件格式（可选），支持``mp3``, ``m4a``。若未设置，则根据文件扩展名对支持的两种音频文件进行格式转换。
+- ``save_dir``: 转换后``.wav``格式文件的存储目录，文件名称与原始音频保持一致。
+
+#### 自定义数据标注
+
+对于不同的应用场景，关键信息的配置多种多样，直接应用通用信息抽取模型的效果可能不够理想。这时可以标注少量场景相关的数据，利用few-shot learning技术来改进特定场景下的信息抽取效果。
+
+自定义数据的格式应与[语音报销工单数据](https://paddlenlp.bj.bcebos.com/datasets/erniekit/speech-cmd-analysis/audio-expense-account.jsonl)相同，划分为训练集``train.txt``和验证集``dev.txt``，保存在``./data/``目录下。
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: 这里的逻辑有点奇怪，看起来下面是有脚本进行数据集划分的

File: applications/speech-cmd-analysis/README.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    --negative_ratio 5 \
+    --splits 0.6 0.8
+```
+
+可配置参数包括
+
+- ``input_file``: 原始数据文件名。文件内容应与[语音报销工单数据](https://paddlenlp.bj.bcebos.com/datasets/erniekit/speech-cmd-analysis/audio-expense-account.jsonl)的格式一致。
+- ``save_dir``: 训练数据的保存目录。若``splits``为空，则数据存储在``processed_data.txt``文件，若``splits``设置为长度为2的列表，则数据存储在目录下的``train.txt``、``dev.txt``、``test.txt``文件。
+- ``negative_ratio``: 负样本与正样本的比例。使用负样本策略可提升模型效果，负样本数量 = negative_ratio * 正样本数量。
+- ``splits``: 划分数据集时训练集、验证集所占的比例。默认为[0.6, 0.8]表示按照``6:2:2``的比例将数据划分为训练集、验证集和测试集。
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: 这块为了体现few-shot效果 是不是可以把训练少一点

File: applications/speech-cmd-analysis/README.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+#### 预训练模型参数
+
+下载预训练好的[UIE模型](https://bj.bcebos.com/paddlenlp/taskflow/information_extraction/uie_base/model_state.pdparams)，放在`./uie_model/`目录下。
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md:143]: 模型这样需要依赖工具自动下载，后续可以修改一下

File: applications/speech-cmd-analysis/README.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+- ``num_epochs``: 训练轮数。
+- ``logging_steps``: 日志打印间隔的step数。
+- ``valid_steps``: 在验证集上进行评估间隔的step数。
+- ``device``: 模型训练使用的设备，可选cpu或gpu。
+- ``seed``: 随机数种子，用于训练过程复现。
+
+
+## 5. 模型预测
+
+在语音报销工单信息录入场景下，按照第3节中的要求准备好音频文件，执行语音指令解析脚本即可抽取报销需要的``时间``、``出发地``、``目的地``和``费用``字段。具体命令如下
```
Comment (wawltor) [inline applications/speech-cmd-analysis/README.md]: 这里需要提示一下，few-shot样本数据字段需要和scheam绑定，提示用户不要轻易改变``时间``、``出发地``、``目的地``和``费用``字段

File: applications/speech-cmd-analysis/finetune.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    rank = paddle.distributed.get_rank()
+    if paddle.distributed.get_world_size() > 1:
+        paddle.distributed.init_parallel_env()
+
+    set_seed(args.seed)
+
+    tokenizer = AutoTokenizer.from_pretrained('ernie-1.0')
+    model = UIE()
+
+    if args.init_from_ckpt and os.path.isfile(args.init_from_ckpt):
```
Comment (wawltor) [inline applications/speech-cmd-analysis/finetune.py]: 这块还是要判断一下模型是否存在，不存在提示用户下载模型

File: applications/speech-cmd-analysis/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        help='The path to save the recognised text and schemas.')
+    args = parser.parse_args()
+
+    if os.path.isfile(args.audio_file):
+        audios = [args.audio_file]
+    elif os.path.isdir(args.audio_file):
+        audios = [x for x in os.listdir(args.audio_file)]
+        audios = [os.path.join(args.audio_file, x) for x in audios]
+    else:
+        raise IOError('%s is neither valid path nor file!' % args.audio_file)
```
Comment (wawltor) [inline applications/speech-cmd-analysis/pipeline.py]: 这里应该不是IO错误吧，普通的异常就行

File: applications/speech-cmd-analysis/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    audios = [x for x in audios if x.endswith('.wav')]
+    if len(audios) == 0:
+        raise IOError('No valid .wav file! Please check %s.' % args.audio_file)
+
```
Comment (wawltor) [inline applications/speech-cmd-analysis/pipeline.py:57]: 同上

File: applications/speech-cmd-analysis/preprocess.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+from utils import set_seed, save_examples, convert_data_examples
+
+
+def convert_data_file(data_file,
+                      save_dir,
+                      splits=[0.8, 0.9],
+                      negative_ratio=1,
+                      is_shuffle=True):
+    """
+        @Description: Consvert file to data format which is suitable to input to this Application.
```
Comment (wawltor) [inline applications/speech-cmd-analysis/preprocess.py]: 这个看起来不是PaddleNLP的文档风格

File: applications/speech-cmd-analysis/utils.py [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+        audio_file (str):
+            Audio file of Mandarin with sampling rate 16000.
+        audio_format (str):
+            The file extension of audio_file, 'wav' by default. 
+    """
+    # Configurations.
+    TOKEN_URL = 'http://aip.baidubce.com/oauth/2.0/token'
+    ASR_URL = 'http://vop.baidu.com/server_api'
+    SCOPE = 'audio_voice_assistant_get'
+    API_KEY = 'vMkN1f7KlL3jornZnQGufBt4'
```
Comment (wawltor) [inline applications/speech-cmd-analysis/utils.py]: 这块的AK SK 是可以公开的吗？

File: applications/speech-cmd-analysis/utils.py [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+def evaluate(model, data_loader):
+    """
+    Given a dataset, it evals model and computes the metric.
+    Args:
+        model(obj:`paddle.nn.Layer`): A model to classify texts.
+        data_loader(obj:`paddle.io.DataLoader`): The dataset loader which generates batches.
+    """
+    model.eval()
+    num_correct = 0
+    num_infer = 0
```
Comment (wawltor) [inline applications/speech-cmd-analysis/utils.py:120]: 部分函数需要迁移

File: applications/speech-cmd-analysis/utils.py (hunk 12 above)
Comment (wawltor) [reply applications/speech-cmd-analysis/utils.py]: 如果免费账号到期了，是不是也要引导一下用户注册百度大脑的账号

File: applications/speech-cmd-analysis/README.md (hunk 1 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md]: done.

File: applications/speech-cmd-analysis/README.md (hunk 2 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md]: 已删除。

File: applications/speech-cmd-analysis/README.md (hunk 3 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md]: done

File: applications/speech-cmd-analysis/README.md (hunk 4 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md]: 数据集划分的叙述已删除。

File: applications/speech-cmd-analysis/pipeline.py (hunk 9 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/pipeline.py]: 已改为Exception。

File: applications/speech-cmd-analysis/pipeline.py (hunk 10 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/pipeline.py:57]: 同上

File: applications/speech-cmd-analysis/preprocess.py (hunk 11 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/preprocess.py]: 已改为PaddleNLP风格的注释。

File: applications/speech-cmd-analysis/utils.py (hunk 12 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/utils.py]: 已改为配置参数。在文档中增加了AK、SK申请教程。

File: applications/speech-cmd-analysis/utils.py (hunk 13 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/utils.py:120]: TODO

File: applications/speech-cmd-analysis/README.md (hunk 7 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md]: done

File: applications/speech-cmd-analysis/README.md (hunk 6 above)
Comment (LemonNoel) [reply applications/speech-cmd-analysis/README.md:143]: 已修改，默认加载uie-base模型来初始化参数。

**Review by wawltor — APPROVED** [review]
LGTM
